---
title: "Assignment 6 RNAseqPart 1"
output: html_document
---

__Name:__ Eliza Tsang
__Student ID:__ 999606858

```{r}
library(rtracklayer)
gff <- import.gff("Brapa_reference/Brapa_gene_v1.5.gff")
summary(gff)

#create a column "gene_id" that contains the gene name for every entry
gff$gene_id <- ifelse(is.na(gff$ID),gff$Parent,gff$ID)

export(gff,"Brapa_reference/Brapa_gene_v1.5.gtf",format="gtf")
```

```{r}
library(Rsubread)
readCounts <- featureCounts(
  files=c("../Assignment_5_Tsang.Eliza/tophat_out-IMB211_All_A01_INTERNODE.fq/accepted_hits_A01.bam",
    "../Assignment_5_Tsang.Eliza/tophat_out-R500_All_A01_INTERNODE.fq/accepted_hits_A01.bam"),
  annot.ext="Brapa_reference/Brapa_gene_v1.5.gtf", 
  isGTFAnnotationFile=TRUE,
  GTF.featureType="CDS", # This depends on GTF file.  Often it would be "exon"
  GTF.attrType="gene_id"
  )

```

__Exercise 1__  
**Read the help file for feature counts.  Be sure to look at the section "Value" where it describes the output.**

```{r}
help("featureCounts")
```

__a)__ **Provide a line of code that displays the counts of the first 6 genes.  (It is not very interesting because the first genes in the file are on chromosome A03 (strange numbering...) and our bam file only has counts from A01...  )**

```{r}
summary(readCounts)
head(readCounts$annotation)
```   

__b__ **The gene `Bra011030` is on chromosome A01.  What are its read counts in the two files?**

```{r}
head(readCounts$counts['Bra011030',])
```

__c)__ **What percentage of reads (from each file) were assigned to a gene?  What percentage were unassigned because they were not located in a gene (aka "Feature")?**

 File 1: ../Assignment_5_Tsang.Eliza/tophat_out-IMB211_All_A01 ... ||
    
    Total reads : 1974510                                               
    
    Successfully assigned reads : 1601807 (81.1%) 
    
 File 2: ../Assignment_5_Tsang.Eliza/tophat_out-R500_All_A01_I ... ||
   
    Total reads : 2191853                                                 
    Successfully assigned reads : 1776266 (81.0%)    
    
Features : 206584 (4.96%); 
Meta-features : 41020 (0.98%)

__d)__ **What are 2 possible reasons why there are reads that cannot be assigned to a gene? **

Some reads could have overlap with others, or there's ambiguity in mapping the reads.
The read could have not been from any of the genes due to contamination.

__Exercise 2__  

__a)__**Create a new object in R called `counts.data` with either the leaf data or the internode data.  **

```{r}
counts.data <-read.csv("gh_internode_counts.tsv",sep = "\t" ) #separate by tabs

```

__b)__**Check to make sure that the data looks as expected.  (What do you expect and how do you confirm?  Show your commands and output.)**

I expected a table with the genes and their reads. At first the data didn't look correct because I didn't separate the columns (using the sep="" argument). [This is shown in the command under 2a.] After adding that, the table looked better but there was still the top row labelled with “*”. I removed them since it did not map to a gene. 

```{r}
counts.data <- counts.data[rownames(counts.data)!="*",]
counts.data[is.na(counts.data)] <- 0
```

__Exercise 3__  
**The column names are too long.  Use the `sub()` command to remove the ".1_matched.merged.fq.bam" suffix from each column name.  Although it doesn't matter in this case, using the argument "fixed=TRUE" is a good idea because "." is a wildcard character.**

```{r}
help("sub")

colnames(counts.data)<- sub(".1_matched.merged.fq.bam" , "", colnames(counts.data), fixed=TRUE) 
head(counts.data)
names(counts.data)
```


__Exercise 4__  
__a.__ Make a histogram of counts for each of the samples.  

```{r}
library(ggplot2)
library(reshape2)
counts.data.melt<- melt(as.matrix(counts.data),value.name="counts", varnames=c("geneID","gr_trt"))
head(counts.data.melt)

ggplot(counts.data.melt,aes(x=counts,fill=gr_trt)) + geom_histogram() + scale_x_log10() +facet_wrap(facets= ~ gr_trt, ncol=3)

```
__b.__ **Is the data normally distributed?  Apply an appropriate transformation if needed and make a new set of histograms. ** 

Before transforming the data with scale_x_log10(), all the data seemed to lean left, with a strong right skew. 

After transforming, we can see clearly that the data is not normally distributed; it's slightly skewed right overall as a result of the left half. Rather than a smooth curve, the left half has several peaks decreasing before rising into the bell.

```{r}
counts.data <- counts.data[rowSums(counts.data > 10) >= 3,]
```

__Exercise 5:__  
**We expect that read counts, especially from biological replicates, will be highly correlated.  Check to see if this is the case using the `pairs()` function and the `cor()` function.  Explain what each of these functions does and comment on the results.  __Important Hint:__ _`pairs` is slow on the full dataset.  Try it on the first 1,000 genes.  Do you need to transform to make the pairs output more meaningful?_**

```{r}
pairs(counts.data[1:1000,])
cor(counts.data)
```
The pairs() command creates a plot while the cor() command lists numerical values. Because I used a smaller sample size to map with pairs, it seems fine without a transformation.

```{r}
sample.description <- data.frame(
  sample=colnames(counts.data),
  gt=regmatches(colnames(counts.data),regexpr("R500|IMB211",colnames(counts.data))),
  
  
  trt=regmatches(colnames(counts.data),regexpr("NDP|DP",colnames(counts.data)))
  )

sample.description$group <- paste(sample.description$gt,sample.description$trt,sep="_")

sample.description$trt <- relevel(sample.description$trt,ref="NDP")

head(sample.description)
summary(sample.description)
```

```{r}
library(edgeR)
dge.data <- DGEList(counts=counts.data, group=sample.description$group)
dim(dge.data) 
dge.data <- calcNormFactors(dge.data, method = "TMM")
dge.data$samples
plotMDS(dge.data, method = "bcv") 
```
__Exercise 6__  
**Discuss the MDS plot.  Does it give you confidence in the experiment or cause concern?**

Each group of internodes reside in a specific area on the plot: IMB211_DP in the top left, IMB211_NDP in the bottom left, R500_DP in the top right, and R500_NDP in the bottom right distances. Each group of internodes have similar biological coefficients of variance with each other. 
Similar internodes are plotted around the same distance (for example, all the R500_NDP internodes are clustered in the same area), so it seems to make sense, providing some confidence in the experiement. 

__Exercise 7__  
To get a graphical idea for what the normalization does, make box plots of the count data for each sample before and after normalization.  Discuss the effect of normalization.

```{r}
counts.data.normal <- cpm(dge.data)

library(reshape2)
counts.data.normal.melt<- melt(as.matrix(counts.data.normal),value.name="counts", varnames=c("geneID","gr_trt"))
head(counts.data.normal.melt)
summary(counts.data.normal.melt)

ggplot(counts.data.normal.melt,aes(x=gr_trt, y=counts, fill=gr_trt)) + geom_boxplot()
```



```{r}
counts.data.log <- log2(counts.data + 1)
head(counts.data.log)
counts.data.log.melt<- melt(as.matrix(counts.data.log),value.name="counts", varnames=c("geneID","gr_trt"))

ggplot(counts.data.log.melt,aes(x=gr_trt, y=counts, fill=gr_trt)) + geom_boxplot() 
```
Transforming the plot normalizes the data, making it more readable and consistent across the data.

__Exercise 8__  
```{r}
design <- model.matrix(~gt+trt,data = sample.description)
rownames(design) <- sample.description$sample
head(design)
summary(design)

dge.data <- estimateGLMCommonDisp(dge.data,design,verbose = TRUE)
dge.data <- estimateGLMTrendedDisp(dge.data,design)
dge.data <- estimateGLMTagwiseDisp(dge.data,design)

plotBCV(dge.data)
```

```{r}
fit <- glmFit(dge.data, design)
gt.lrt <- glmLRT(fit,coef = "gtR500")

topTags(gt.lrt)
summary(decideTestsDGE(gt.lrt,p.value=0.01))

DEgene.gt <- topTags(gt.lrt,n = Inf)$table[topTags(gt.lrt,n = Inf)$table$FDR<0.01,]
write.csv(DEgene.gt,"DEgenes.gt.csv")

plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  require(reshape2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- merge(tmp.data,sample.description,by.x="row.names",by.y="sample")
  tmp.data <- melt(tmp.data,value.name="log2_cpm",variable.name="gene")
  pl <- ggplot(tmp.data,aes(x=gt,y=log2_cpm,fill=trt))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}

plotDE("Bra009785",dge.data,sample.description)
plotDE(rownames(DEgene.gt)[1:9],dge.data,sample.description)
```

**a)  Find all genes differentially expressed in response to the DP treatment (at a FDR < 0.01).**

```{r}
fit2 <- glmFit(dge.data, design)
trt.lrt <- glmLRT(fit2,coef = "trtDP")

topTags(trt.lrt)
summary(decideTestsDGE(trt.lrt,p.value=0.01))

DEgene.trt <- topTags(trt.lrt,n = Inf)$table[topTags(trt.lrt,n = Inf)$table$FDR<0.01,]
write.csv(DEgene.trt,"DEgenes.trt.csv")
head(DEgene.trt)
summary(DEgene.trt)
```

__b)__  **How many genes are differentially expressed?**

619 observations/genes

__c)__  **Make a plot of the top 9**
```{r}
plotDE(rownames(DEgene.trt)[1:9],dge.data,sample.description)
```


__Exercise 9__:  

```{r}
design.interaction <- model.matrix(~gt*trt,data = sample.description)
rownames(design.interaction) <- sample.description$sample
summary(design.interaction)
head(design.interaction)
```
__a)__ **Repeat the dispersion estimates and model fit but with the new model.  Show code. **

```{r}
dge.data2 <- estimateGLMCommonDisp(dge.data,design.interaction,verbose = TRUE)
dge.data2 <- estimateGLMTrendedDisp(dge.data,design.interaction)
dge.data2 <- estimateGLMTagwiseDisp(dge.data,design.interaction)

plotBCV(dge.data2)

fit3 <- glmFit(dge.data2, design.interaction)
trt.lrt2 <- glmLRT(fit3,coef = "trtDP")

topTags(trt.lrt2)
summary(decideTestsDGE(trt.lrt2,p.value=0.01))
```
__b)__ **How many genes show a significantly different response to treatment in IMB211 as compared to R500?  Save these genes to a file.**  

```{r}
DEgene.trt.interaction <- topTags(trt.lrt2,n = Inf)$table[topTags(trt.lrt2,n = Inf)$table$FDR<0.01,]
write.csv(DEgene.trt.interaction,"DEgenes.trt.interaction.csv")
head(DEgene.trt.interaction)
summary(DEgene.trt.interaction)
```

1214 observations/genes

__c)__ **Make a plot of the top 9 genes that have a significantly different response to treatment in IMB211 as compared to R500.**  

```{r}
plotDE(rownames(DEgene.trt.interaction)[1:9],dge.data2,sample.description)
```